command/agent/acl.go

```diff
@@ -0,0 +1,452 @@
+	ETag string
+}
+
+// aclManager is used by the agent to keep track of state related to ACLs,
+// including caching tokens from the servers. This has some internal state that
+// we don't want to dump into the agent itself.
+type aclManager struct {
+	// acls is a cache mapping ACL tokens to compiled policies.
+	acls *lru.TwoQueueCache
+
+	// master is the ACL to use when the agent master token is supplied.
+	// This may be nil if that option isn't set in the agent config.
+	master acl.ACL
+
+	// down is the ACL to use when the servers are down. This may be nil
+	// which means to try and use the cached policy if there is one (or
+	// deny if there isn't a policy in the cache).
+	down acl.ACL
+
+	// disabled is used to keep track of feedback from the servers that ACLs
+	// are disabled. If the manager discovers that ACLs are disabled, this
+	// will be set to the next time we should check to see if they have been
+	// enabled. This helps cut useless traffic, but allows us to turn on ACL
+	// support at the servers without having to restart the whole cluster.
+	disabled     time.Time
+	disabledLock sync.RWMutex
+}
+
+// newACLManager returns an ACL manager based on the given config.
+func newACLManager(config *Config) (*aclManager, error) {
+	// Set up the cache from ID to ACL (we don't cache policies like the
+	// servers; only one level).
+	acls, err := lru.New2Q(aclCacheSize)
+	if err != nil {
+		return nil, err
+	}
+
+	// If an agent master token is configured, build a policy and ACL for
+	// it, otherwise leave it nil.
+	var master acl.ACL
+	if len(config.ACLAgentMasterToken) > 0 {
+		policy := &acl.Policy{
+			Agents: []*acl.AgentPolicy{
+				&acl.AgentPolicy{
+					Node:   config.NodeName,
+					Policy: acl.PolicyWrite,
+				},
+			},
+		}
+		acl, err := acl.New(acl.DenyAll(), policy)
+		if err != nil {
+			return nil, err
+		}
+		master = acl
+	}
+
+	var down acl.ACL
+	switch config.ACLDownPolicy {
+	case "allow":
+		down = acl.AllowAll()
+	case "deny":
+		down = acl.DenyAll()
+	case "extend-cache":
+		// Leave the down policy as nil to signal this.
+	default:
+		return nil, fmt.Errorf("invalid ACL down policy %q", config.ACLDownPolicy)
+	}
+
+	// Give back a manager.
+	return &aclManager{
+		acls:   acls,
+		master: master,
+		down:   down,
+	}, nil
+}
+
+// isDisabled returns true if the manager has discovered that ACLs are disabled
+// on the servers.
+func (m *aclManager) isDisabled() bool {
+	m.disabledLock.RLock()
+	defer m.disabledLock.RUnlock()
+	return time.Now().Before(m.disabled)
+}
+
+// lookupACL attempts to locate the compiled policy associated with the given
+// token. The agent may be used to perform RPC calls to the servers to fetch
+// policies that aren't in the cache.
+func (m *aclManager) lookupACL(agent *Agent, id string) (acl.ACL, error) {
+	// Handle some special cases for the ID.
+	if len(id) == 0 {
+		id = anonymousToken
+	} else if acl.RootACL(id) != nil {
+		return nil, errors.New(rootDenied)
```

Comment:
I'm not sure what's being checked here - it seems like this just prevents passing tokens called "allow", "deny", or "manage"?